hasher.go

```diff
@@ -86,16 +99,25 @@ func (n *Hasher) Write(data []byte) (int, error) {
 
 // Sum computes the hash. Does not append the given suffix, violating the
 // interface.
+// It may panic if the data being hashed is invalid. This should never happen since the Write method refuses an invalid data and errors out.
 func (n *Hasher) Sum([]byte) []byte {
 	switch n.tp {
 	case LeafPrefix:
-		return n.HashLeaf(n.data)
+		res, err := n.HashLeaf(n.data)
+		if err != nil {
+			panic(err) // this should never happen since the data is already validated in the Write method
+		}
```

Comment:
[question]
could this still be hit when verifying a proof since we're not calling the write method ahead of time?

Reply:
If the `Write` method is not called prior to the `Sum`, then the code panics due to the absence of the `tp` field of the `Hasher`. More details follow:
For the `Sum` to get through, the `tp` field of the hasher should be set to one of the `LeafPrefix` or `NodePrefix` based on which the hasher determines whether the input data belongs to a leaf or a non-leaf node. Without making call to the `Write` method, the `Hasher.tp` field remains unset, hence in the mentioned switch statement, none of the `LeafPrefix` and `NodePrefix` cases are true, and the code goes straight to the `default` part which panics due to unset `tp` field.

Reply:
Another point is that, it is not possible to call `Sum` without making a call to the `Write` method, as it corresponds to doing nothing (the `Hasher` state is empty). Doing so always results in panic due to undefined state https://github.com/celestiaorg/nmt/blob/85ec3c9e7c593bf95f48c8cf7665097cc5f00e26/hasher.go#L121.